util/src/overlaydb.rs

```diff
@@ -36,344 +43,236 @@ use kvdb::{Database, DBTransaction};
 ///
 /// `lookup()` and `contains()` maintain normal behaviour - all `insert()` and `remove()`
 /// queries have an immediate effect in terms of these functions.
+///
+/// The total amount of insertions and removals of a single value must add up to
+/// either
+///   - A single insertion
+///   - A single deletion
+///   - No operation.
+///
+/// Additionally, it is an error to insert a value which already exists in the backing
```

Comment:
the actual "user" of this struct doesn't care about the backing DB as it's an implementation issue.

so as i understand it, the semantics of this struct are:
- `insert`s and `delete`s may happen multiple times each for the same key;
- at the point of any `commit`, the total historic insertions (+1) and deletions (-1) must sum to either ONE or ZERO;
- this struct `contain`s a key which may be duly dereferenced iff the total historical insertions (+1) and deletions (-1) sum to ONE;
- a `revert`, or the cessation of program operation, ensures the state (aka "total history") of this struct is as of the `commit` immediately prior.